readme.md
# canvas-gauge-card
![Build](https://github.com/custom-cards/canvas-gauge-card/workflows/Build/badge.svg)

This card is basically deprecated and do not expect maintainence. 

## Development
I recommend use vscode devcontainers. do 
```shell
npm install
npm run build
```
Inside container you can build, start HA using vscode tasks.

This card allows you to use the awesome gauges at https://canvas-gauges.com/ in your lovelace GUI. This is still a very early and untested alpha version. Tested only on chrome and safari browsers.

If you want to support the developer, please support the lib devs of the canvas-gauges.

## Screens

Here is a some different screens from my own setup. But the possibilites to make your own look and feel are great.

<a href="docs/screen_1.png" target="_blank"><img src="docs/screen_1.png"  width="254" height="204"/></a>
<a href="docs/screen_2.png" target="_blank"><img src="docs/screen_2.png"  width="250" height="244"/></a>

## Usage

Download through HACS (recommended) or ...

Download the file `canvas-gauge-card.js` from the [**latest release**](https://github.com/custom-cards/canvas-gauge-card/releases/latest) to your home assistant config. Add a module link to the `canvas-gauge-card.js` file. The link needs to be the same as where you put it.
Example:

```yaml
resources:
  - url: /local/custom_cards/canvas-gauge-card.js
    type: module
```

Configure the card properties.

Skip `name`property if you dont want the name with the shadow area to show.

Use the javascript names of properties from the examples at https://canvas-gauges.com/documentation/examples/. Click on an example that you like, check the JS version and copy the properties to the lovelace yaml file. Just remove the ',' after copy from site.

### Example 1, simple half gauge

<img src="docs/screen_sample2.png"  width="230" height="130"/>

**ui-lovelace.yaml:**

_Notice the differences in `card_height` and `gauge/height` to get correct half size for full circle, set both to same._

```yaml
- type: custom:canvas-gauge-card
  entity: sensor.processor_use
  card_height: 125
  gauge:
    type: "radial-gauge"
    title: Processor (%)
    width: 220
    height: 220
    borderShadowWidth: 0
    borderOuterWidth: 0
    borderMiddleWidth: 0
    borderInnerWidth: 0
    minValue: 0
    maxValue: 100
    startAngle: 30
    ticksAngle: 180
    valueBox: false
    majorTicks:
      ["0", "10", "20", "30", "40", "50", "60", "70", "80", "90", "100"]
    minorTicks: 2
    strokeTicks: true
    highlights: [{ "from": 80, "to": 100, "color": "rgba(200, 50, 50, .75)" }]
    borders: false
```

### Example 2, simple half gauge with shadow text

<img src="docs/screen_sample1.png"  width="242" height="165"/>

**ui-lovelace.yaml:**

```yaml
- type: custom:canvas-gauge-card
  entity: sensor.processor_use
  name: Processor (%)
  card_height: 145
  shadow_height: 15%
  font_size: 1em
  gauge:
    type: "radial-gauge"
    width: 220
    height: 220
    borderShadowWidth: 0
    borderOuterWidth: 0
    borderMiddleWidth: 0
    borderInnerWidth: 0
    minValue: 0
    maxValue: 100
    startAngle: 90
    ticksAngle: 180
    valueBox: false
    majorTicks:
      ["0", "10", "20", "30", "40", "50", "60", "70", "80", "90", "100"]
    minorTicks: 2
    strokeTicks: true
    highlights: [{ "from": 80, "to": 100, "color": "rgba(200, 50, 50, .75)" }]
    borders: false
```

### Example 3, simple full gauge with shadow text

<img src="docs/screen_sample3.png"  width="234" height="221"/>

**ui-lovelace.yaml:**

```yaml
- type: custom:canvas-gauge-card
  entity: sensor.processor_use
  name: Processor (%)
  card_height: 210
  shadow_height: 12%
  font_size: 1em
  gauge:
    type: "radial-gauge"
    width: 220
    height: 220
    borderShadowWidth: 0
    borderOuterWidth: 0
    borderMiddleWidth: 0
    borderInnerWidth: 0
    minValue: 0
    maxValue: 100
    startAngle: 40
    ticksAngle: 280
    valueBox: false
    units: "%"
    majorTicks:
      ["0", "10", "20", "30", "40", "50", "60", "70", "80", "90", "100"]
    minorTicks: 2
    strokeTicks: true
    highlights: [{ "from": 80, "to": 100, "color": "rgba(200, 50, 50, .75)" }]
    borders: false
```

### Example 4, 4 gauges in a row in a horizontal-stack

<img src="docs/screen_sample4.png"  width="505" height="95"/>
This example shows gauges with or without text. This text are placed below the actual gauge this time.

**No text version**

```yaml
---
- type: horizontal-stack
  cards:
    - type: custom:canvas-gauge-card
      entity: sensor.load_1m
      card_height: 62
      background_color: "#FFF"
      gauge:
        type: "radial-gauge"
        borderShadowWidth: 0
        borderOuterWidth: 0
        borderMiddleWidth: 0
        borderInnerWidth: 0
        width: 110
        height: 110
        minValue: 0
        maxValue: 100
        startAngle: 90
        ticksAngle: 180
        valueBox: false
        majorTicks:
          ["0", "10", "20", "30", "40", "50", "60", "70", "80", "90", "100"]
        minorTicks: 2
        strokeTicks: true
        highlights:
          [{ "from": 80, "to": 100, "color": "rgba(200, 50, 50, .75)" }]
        colorPlate: "#ddd"
        borders: false
        needleType: "arrow"
        needleWidth: 2
        needleCircleSize: 7
        needleCircleOuter: true
        needleCircleInner: false
        animationDuration: 1500
        animationRule: "linear"
```

**With text version**

```yaml
---
- type: horizontal-stack
  cards:
    - type: custom:canvas-gauge-card
      entity: sensor.processor_use
      card_height: 62
      name: "Processor use"
      shadow_height: "25%"
      font_size: 0.9em
      shadow_bottom: "20"
      gauge:
        type: "radial-gauge"
        width: 110
        height: 110
        borderShadowWidth: 0
        borderOuterWidth: 0
        borderMiddleWidth: 0
        borderInnerWidth: 0
        minValue: 0
        maxValue: 100
        startAngle: 90
        ticksAngle: 180
        valueBox: false
        majorTicks:
          ["0", "10", "20", "30", "40", "50", "60", "70", "80", "90", "100"]
        minorTicks: 2
        strokeTicks: true
        highlights:
          [{ "from": 80, "to": 100, "color": "rgba(200, 50, 50, .75)" }]
        colorPlate: "#ddd"
        borders: false
        needleType: "arrow"
        needleWidth: 2
        needleCircleSize: 7
        needleCircleOuter: true
        needleCircleInner: false
        animationDuration: 1500
        animationRule: "linear"
```

### Properties

Some of the properties that could be set. _italic_ is not mandatory.

| Property           | Description                                                                                                                        |
| :----------------- | ---------------------------------------------------------------------------------------------------------------------------------- |
| entity             | your sensor                                                                                                                        |
| name               | shows the name in shadow remove to hide it                                                                                         |
| type               | `"radial-gauge"` or `"linear-gauge"`                                                                                               |
| dropshadow         | true to show dropshadow, false to hide                                                                                             |
| width              | width of the gauge                                                                                                                 |
| height             | height of the gauge                                                                                                                |
| _background_color_ | sets gauge background color, transparent default if not set                                                                        |
| _card_height_      | the actual height of the card, set to smaller value than gauge height if using a half guage. Not use if using a full circle gauge. |
| _font_size_        | size of name, leave out it will be dynamic                                                                                         |
| _shadow_height_    | xx% of total height is shadow height                                                                                               |
| _shadow_bottom_    | how far below the gauge in pixels the shadow should apear                                                                          |

For a complete documentation of available properties, please see https://canvas-gauges.com/documentation/user-guide/configuration

## Repos

- [helto4real, Original cards](https://github.com/helto4real/lovelace-custom-cards)
- [helto4real, Tomas Hellströms, Hass config using the card](https://github.com/helto4real/hassio)

## Credits

- [Mikhus, canvas guages](https://github.com/Mikhus/canvas-gauges)


RELEASENOTES.md
# 0.0.3

This version fixes problems with version 106.x of Home Assistant. If you have problems with ugly borders please use. There is a bug in the framework that shows borders even if "border: false" is set.

```yaml
gauge:
  borderShadowWidth: 0
  borderOuterWidth: 0
  borderMiddleWidth: 0
  borderInnerWidth: 0
```

# 0.0.5

- Support to set `background_color`, see readme.md.
- Build system, now builds to one file

# 0.0.6

Update to latest guage-card package to 2.1.7

# 0.7.0
Fix animation issues

# 0.8.0

- Dropshadow is now optional
- Removed dependency on custom-card-helpers
- Added devconainer that let you run HA and debug card

# 0.9.0

- Using class instead of id in css
- shadow_height now not visible if not configured.


src/types.ts
// Attribution: to all devs at https://github.com/custom-cards/custom-card-helpers that let me use HomeAssistant object and its deps. Thanks alot!

import {
  HassEntities,
  HassConfig,
  Auth,
  Connection,
  MessageBase,
  HassServices,
  HassServiceTarget,
} from "home-assistant-js-websocket";

export interface CanvasGuageCardConfig {
  entity?: string;
  name?: string;
  show_warning?: boolean;
  show_error?: boolean;
  gauge?: any;
  font_size?: string;
  card_width?: string;
  card_height?: string;
  card_top?: string;
  card_left?: string;
  dropshadow?: boolean;

  shadow_height?: string;
  shadow_bottom?: string;
  background_color?: string;
}

export interface ServiceCallRequest {
  domain: string;
  service: string;
  serviceData?: Record<string, any>;
  target?: HassServiceTarget;
}
export enum NumberFormat {
  language = "language",
  system = "system",
  comma_decimal = "comma_decimal",
  decimal_comma = "decimal_comma",
  space_comma = "space_comma",
  none = "none",
}

export enum TimeFormat {
  language = "language",
  system = "system",
  am_pm = "12",
  twenty_four = "24",
}

export interface FrontendLocaleData {
  language: string;
  number_format: NumberFormat;
  time_format: TimeFormat;
}

export interface Theme {
  // Incomplete
  "primary-color": string;
  "text-primary-color": string;
  "accent-color": string;
}

export interface Themes {
  default_theme: string;
  themes: { [key: string]: Theme };
}

export interface Panel {
  component_name: string;
  config: { [key: string]: any } | null;
  icon: string | null;
  title: string | null;
  url_path: string;
}

export interface Panels {
  [name: string]: Panel;
}

export interface Resources {
  [language: string]: { [key: string]: string };
}

export interface Translation {
  nativeName: string;
  isRTL: boolean;
  fingerprints: { [fragment: string]: string };
}

export interface ServiceCallRequest {
  domain: string;
  service: string;
  serviceData?: Record<string, any>;
  target?: HassServiceTarget;
}

export type LocalizeFunc = (key: string, ...args: any[]) => string;

export interface Credential {
  auth_provider_type: string;
  auth_provider_id: string;
}

export interface MFAModule {
  id: string;
  name: string;
  enabled: boolean;
}

export interface CurrentUser {
  id: string;
  is_owner: boolean;
  is_admin: boolean;
  name: string;
  credentials: Credential[];
  mfa_modules: MFAModule[];
}

export interface HomeAssistant {
  auth: Auth;
  connection: Connection;
  connected: boolean;
  states: HassEntities;
  services: HassServices;
  config: HassConfig;
  themes: Themes;
  selectedTheme?: string | null;
  panels: Panels;
  panelUrl: string;

  language: string;
  locale: FrontendLocaleData;
  // local stored language, keep that name for backward compability
  selectedLanguage: string | null;
  resources: Resources;
  localize: LocalizeFunc;
  translationMetadata: {
    fragments: string[];
    translations: {
      [lang: string]: Translation;
    };
  };

  dockedSidebar: boolean;
  moreInfoEntityId: string;
  user: CurrentUser;
  callService: (
    domain: ServiceCallRequest["domain"],
    service: ServiceCallRequest["service"],
    serviceData?: ServiceCallRequest["serviceData"],
    target?: ServiceCallRequest["target"]
  ) => Promise<void>;
  callApi: <T>(
    method: "GET" | "POST" | "PUT" | "DELETE",
    path: string,
    parameters?: { [key: string]: any }
  ) => Promise<T>;
  fetchWithAuth: (
    path: string,
    init?: { [key: string]: any }
  ) => Promise<Response>;
  sendWS: (msg: MessageBase) => Promise<void>;
  callWS: <T>(msg: MessageBase) => Promise<T>;
}

src/canvas-gauge-card.ts
/*!
 * The MIT License (MIT)
 *
 * Copyright (c) 2017 Tomas Hellström (@helto4real) (https://github.com/custom-cards/canvas-gauge-card)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * `canvas-gauge-card`
 * Lovelace element for displaying the canvas gauges at
 * https://canvas-gauges.com/
 *
 * If you like the gauges please support the original gauge devs.
 *
 * Usage:
 *  - see https://github.com/custom-cards/canvas-gauge-card
 */

import {
  LitElement,
  html,
  customElement,
  property,
  CSSResult,
  TemplateResult,
  css,
  query,
  PropertyValues,
} from "lit-element";

import Gauge from "canvas-gauges";

import { HomeAssistant, CanvasGuageCardConfig } from "./types.js";
import { CARD_VERSION } from "./const.js";
import { localize } from "./localize/localize.js";

// Check if config or Entity changed
export function hasConfigOrEntityChanged(
  element: any,
  changedProps: PropertyValues,
  forceUpdate: Boolean,
): boolean {
  if (changedProps.has('config') || forceUpdate) {
    return true;
  }

  if (element.config!.entity) {
    const oldHass = changedProps.get('hass') as HomeAssistant | undefined;
    if (oldHass) {
      return (
        oldHass.states[element.config!.entity]
        !== element.hass!.states[element.config!.entity]
      );
    }
    return true;
  } else {
    return false;
  }
}

/* eslint no-console: 0 */
console.info(
  `%c  CANVAS-GAUGE-CARD \n%c  ${localize(
    "common.version"
  )} ${CARD_VERSION}    `,
  "color: orange; font-weight: bold; background: black",
  "color: white; font-weight: bold; background: dimgray"
);

(window as any).customCards = (window as any).customCards || [];
(window as any).customCards.push({
  type: "canvas-gauge-card",
  name: "Canvas Gauge Card",
  description: "A canvas gauge card from http://canvas-gauges.com/",
});

@customElement("canvas-gauge-card")
export class CanvasGaugeCard extends LitElement {
  public static getStubConfig(): object {
    return {};
  }

  @property() public hass?: HomeAssistant;
  @property() private _config?: CanvasGuageCardConfig;

  private _gaugeWidth?: string;
  private _gaugeHeight?: string;
  private _shadowHeight?: string;
  private _fontSize?: string;
  private _useDropshadow?: boolean;

  private _gauge?: any;
  private _state?: string;
  private config?: any;

  @query("#canvaselement") private _canvasElement?: any;

  public setConfig(config: CanvasGuageCardConfig): void {
    if (!config || config.show_error) {
      throw new Error(localize("common.invalid_configuration"));
    }
    // Have no ide why I need to do this or the thing bugs in .. but I do
    this.config = config;

    this._config = {
      gauge: config.gauge,
      ...config,
    };

    this._gaugeWidth = config.card_width
      ? config.card_width
      : config.gauge["width"];

    this._gaugeHeight = config.card_height
      ? config.card_height
      : config.gauge["height"];

    this._shadowHeight = config.shadow_height ? config.shadow_height : "0%";
    this._useDropshadow = config.dropshadow ? config.dropshadow : false;

    this._fontSize = config.font_size
      ? config.font_size
      : `calc(${config.gauge["height"]}px/22)`;
  }

  protected shouldUpdate(changedProps: PropertyValues): boolean {
    return hasConfigOrEntityChanged(this, changedProps, false);
  }

  // Here we need to refresh the actual gauge after it has rendered
  protected updated(_) {
    if (this._gauge == null) {
      var gauge;
      if (this._config?.gauge.type == "linear-gauge") {
        gauge = new Gauge.LinearGauge({
          renderTo: this._canvasElement,
          height: this._config.gauge["height"],
          width: this._config.gauge["width"],
          value: 0,
        });
      } else if (this._config?.gauge.type == "radial-gauge") {
        gauge = new Gauge.RadialGauge({
          renderTo: this._canvasElement,
          height: this._config.gauge["height"],
          width: this._config.gauge["width"],
          value: 0,
        });
      }

      for (const key in this._config?.gauge) {
        if (this._config?.gauge?.hasOwnProperty(key)) {
          gauge.options[key] = this._config.gauge[key];
        }
      }
      this._gauge = gauge;
    }
    var entityId = this._config?.entity;
    this._state = this.hass?.states[entityId!].state;
    this._gauge["value"] = this._state;
    this._gauge.draw(); // Have to call to redraw canvas
  }
  protected render(): TemplateResult | void {
    if (!this._config || !this.hass) {
      return html``;
    }

    // Check for stateObj or other necessary things and render a warning if missing
    // if provided in the future, this is a good spot doing just that :)
    if (this._config.show_warning) {
      return html`
        <ha-card>
          <div class="warning">${localize("common.show_warning")}</div>
        </ha-card>
      `;
    }
    return html`
      <style>
        :host {
          ${this._useDropshadow ?
        `          box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14),
            0 1px 5px 0 rgba(0, 0, 0, 0.12), 0 3px 1px -2px rgba(0, 0, 0, 0.2);`
        : ` `}
          display: block !important;
          border-radius: 2px !important;
          transition: all 0.3s ease-out !important;
          background-color: ${this._config.background_color
        ? this._config.background_color
        : "transparent"} !important;
        }
        .cardroot {
          width: ${this._gaugeWidth}px;
          height: calc(
            ${this._gaugeHeight}px +
              ${this._config.shadow_bottom ? this._config.shadow_bottom : 0}px
          );
          position: relative;
          margin: auto;
        }
        .container {
          width: ${this._gaugeWidth}px;
          height: ${this._gaugeHeight}px;
          position: relative;
          top: 0px;
          overflow: hidden;
          text-align: center;
        }
        .innercontainer {
          position: relative;
          top: ${this._config.card_top ? this._config.card_top : 0};
          left: ${this._config.card_left ? this._config.card_left : 0};
        }
        .shadow {
          visible: ${this._shadowHeight == "0%" ? `none` : `block`};
          width: 100%;
          height: ${this._shadowHeight};
          left: 0px;
          bottom: 0px;
          background: rgba(0, 0, 0, 0.5);
          position: absolute;
        }
        .state {
          position: relative;
          float: left;
          top: 50%;
          left: 50%;
          color: white;
          font-size: 100%;
          transform: translate(-50%, -50%);
        }
      </style>
      <div class="cardroot">
        <div
          class="container"
          width=${this._gaugeWidth}
          height=${this._gaugeHeight}
        >
          <div
            class="innercontainer"
            width=${this._gaugeWidth}
            height=${this._gaugeHeight}
            @click=${this.clickHandler}
          >
            <canvas id="canvaselement"> </canvas>
          </div>
        </div>
        <div class="shadow">
          <div class="state" style="font-size: ${this._fontSize}">
            ${this._config.name}
          </div>
        </div>
      </div>
    `;
  }

  static get styles(): CSSResult {
    return css`
    .warning {
  display: block;
  color: black;
  background - color: #fce588;
  padding: 8px;
}
`;
  }
  clickHandler(_) {
    this._fire("hass-more-info", { entityId: this._config?.entity });
  }

  /**
   * Fires the event that opens the enity info
   */
  _fire(type, detail) {
    var event = new Event(type, {
      bubbles: true,
      cancelable: false,
      composed: true,
    }) as any;

    event.detail = detail || {};
    this.shadowRoot?.dispatchEvent(event);
    return event;
  }
}


src/canvas-gauge-card.js
/*!
 * The MIT License (MIT)
 *
 * Copyright (c) 2017 Tomas Hellström (https://github.com/helto4real/lovelace-custom-cards)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import Gauge from "canvas-gauges";

/**
 * `canvas-gauge-card`
 * Lovelace element for displaying the canvas gauges at
 * https://canvas-gauges.com/
 *
 * If you like the gauges please support the original gauge devs.
 *
 * Usage och card:
 *  - Copy this file 'canvas-guage-card.js' and 'gauge.min.js' to same folder in hass.
 *  - Set the ui-lovelace.yaml, Include the canvas-guage-card.js and cofigure at minimum:
    * cards:
        - type: custom:canvas-gauge-card
            entity: sensor.temp_outside
            gauge:
            type: "linear-gauge"
            width: 120
            height: 400
    - Use the javascript properties withouth the ',' to configure properties under 'gauge'
    - Use name property to show the name att bottom of the card
    - See http://sss for docs
 */
class CanvasGaugeCard extends HTMLElement {
  /**
   * Renders the card
   */
  _render() {
    const elemCardRoot = document.createElement("div");
    elemCardRoot.id = "cardroot";
    // Create the container element
    const elemContainer = document.createElement("div");
    elemContainer.id = "container";
    elemContainer.width = this.config.card_width
      ? this.config.card_width
      : this.config.gauge["width"];
    elemContainer.height = this.config.card_height
      ? this.config.card_height
      : this.config.gauge["height"];

    const shadowHeight = this.config.shadow_height
      ? this.config.shadow_height
      : "10%";
    // The styles
    const style = `
            <style>
                :host {
                    box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 1px 5px 0 rgba(0, 0, 0, 0.12), 0 3px 1px -2px rgba(0, 0, 0, 0.2);
                    display: block !important;
                    border-radius: 2px !important;
                    transition: all 0.30s ease-out !important;
                    background-color: ${
                      this.config.background_color
                        ? this.config.background_color
                        : "transparent"
                    } !important;
                }
                #cardroot {
                    width: ${elemContainer.width}px;
                    height: calc(${elemContainer.height}px + ${
      this.config.shadow_bottom ? this.config.shadow_bottom : 0
    }px);
                    position: relative;
                    margin: auto;
                }
                #container {
                    width: ${elemContainer.width}px;
                    height: ${elemContainer.height}px;
                    position: relative;
                    top: 0px;
                    overflow: hidden;
                    text-align: center;
                }
                #innercontainer {
                    position: relative;
                    top: ${this.config.card_top ? this.config.card_top : 0}
                    left: ${this.config.card_left ? this.config.card_left : 0}
                }
                .shadow {
                    width: 100%;
                    height: ${shadowHeight};
                    left: 0px;
                    bottom: 0px;
                    background: rgba(0, 0, 0, 0.5);;
                    position: absolute;
                    }
                #state{
                    position: relative;
                    float: left;
                    top: 50%;
                    left: 50%;
                    color: white;
                    font-size: 100%;
                    transform: translate(-50%, -50%);
                }
            </style>

        `;
    this.shadowRoot.innerHTML = style;

    // We need an inner container so we can hide part of the gauge cause it
    // renders full circle even if only showing half gauge.
    const elemInnerContainer = document.createElement("div");
    elemInnerContainer.id = "innercontainer";

    // canvas element that the gauge will be drawn by the canvas gauge Lib
    const elemCanvas = document.createElement("canvas");
    elemCanvas.width = this.config.gauge["width"];
    elemCanvas.height = this.config.gauge["height"];

    // Have to do this cause some bug in library gauge.min.js dont display the gauge
    // keep this code for a while incase I need to activate it again...
    //var ctx = elemCanvas.getContext('2d');
    //ctx.fillStyle = 'black';
    //ctx.fillRect(0, 0, elemCanvas.width, elemCanvas.height);
    var gauge;
    if (this.config.gauge.type == "linear-gauge") {
      gauge = new Gauge.LinearGauge({
        renderTo: elemCanvas,
        height: elemCanvas.height,
        width: elemCanvas.width,
        value: 0
      });
    } else if (this.config.gauge.type == "radial-gauge") {
      gauge = new Gauge.RadialGauge({
        renderTo: elemCanvas,
        height: elemCanvas.height,
        width: elemCanvas.width,
        value: 0
      });
    }

    for (const key in this.config.gauge) {
      if (this.config.gauge.hasOwnProperty(key)) {
        gauge.options[key] = this.config.gauge[key];
      }
    }

    elemInnerContainer.appendChild(elemCanvas);

    elemContainer.appendChild(elemInnerContainer);
    elemCardRoot.appendChild(elemContainer);
    elemContainer.onclick = this._click.bind(this);
    if (this.config.name) {
      var elemShadow = document.createElement("div");
      elemShadow.className = "shadow";

      var elemState = document.createElement("div");
      elemState.id = "state";
      // Automatic font resize or set one
      var font_size = this.config.font_size
        ? this.config.font_size
        : `calc(${this.config.gauge["height"]}px/22)`;
      elemState.style.fontSize = font_size;
      elemState.innerText = this.config.name;

      elemShadow.appendChild(elemState);
      elemCardRoot.appendChild(elemShadow);
    }

    this.shadowRoot.appendChild(elemCardRoot);
    this._gauge = gauge;
  }

  /**
   * onclick event for card, gets the enity info
   */
  _click() {
    this._fire("hass-more-info", { entityId: this.config.entity });
  }

  /**
   * Fires the event that opens the enity info
   */
  _fire(type, detail) {
    const event = new Event(type, {
      bubbles: true,
      cancelable: false,
      composed: true
    });
    event.detail = detail || {};
    this.shadowRoot.dispatchEvent(event);
    return event;
  }

  set hass(hass) {
    const entityId = this.config.entity;
    this._state = hass.states[entityId].state;
    this._gauge["value"] = this._state;
    this._gauge.draw(); // Have to call to redraw canvas
  }

  setConfig(config) {
    if (!config.entity) {
      throw new Error("You need to define an entity");
    }
    if (!config.gauge) {
      throw new Error("You need to define gauge and default gauge values");
    }
    if (!config.gauge.height) {
      throw new Error("You need to define gauge height");
    }
    if (!config.gauge.width) {
      throw new Error("You need to define gauge width");
    }
    if (!config.gauge.type) {
      throw new Error("You need to define gauge type");
    }
    if (
      !(
        config.gauge.type == "linear-gauge" ||
        config.gauge.type == "radial-gauge"
      )
    ) {
      throw new Error(
        'You need to define gauge type "linear-gauge" or "radial-gauge"'
      );
    }
    this.config = config;

    // Remove, fails from v0.106
    // Fix initbug from the canvas lib that shows borders even if set to false
    // if (typeof config.gauge.borders != typeof undefined && config.gauge.borders === false) {
    //     config.gauge['borderShadowWidth'] = 0;
    //     config.gauge['borderOuterWidth'] = 0;
    //     config.gauge['borderMiddleWidth'] = 0;
    //     config.gauge['borderInnerWidth'] = 0;
    // }
    this._render();
  }

  getCardSize() {
    return 2;
  }

  constructor() {
    super();
    this.attachShadow({ mode: "open" });
  }
}

window.customElements.define("canvas-gauge-card", CanvasGaugeCard);
